### ISSUE TITLE
Building a dll in windows using c-shared buildmode generates a dll with .exe extension

### ISSUE THREAD

**Issue by qmuntal** [goreleaser/goreleaser#1221]
**Describe the bug**

When building a binary using `-buildmode=c-shared` in windows go generates a dll, not an executable, but goreleaser always add the extension `.exe` to all the binaries generated with goos=windows, at least this is what I can see in the [source code](https://github.com/goreleaser/goreleaser/blob/master/internal/pipe/build/build.go#L131).

**To Reproduce**

```yaml
builds:
  - id: "testdll"
    env:
      - CGO_ENABLED=1
      - CC=x86_64-w64-mingw32-gcc
      - CXX=x86_64-w64-mingw32-g++
    goos:
      - windows
    goarch:
      - amd64
    flags: -buildmode=c-shared
    ldflags: -s -w
    main: .
    binary: test.dll
archives:
  - format: binary
    builds:
      - testdll
```

**Expected behavior**

I'm expecting a binary archive named `test.dll`, but I get a `test.dll.exe` archive.

**Environment (please complete the following information):**

- OS: windows
- OS version: windows 10 x64
- GoReleaser Version: 0.120.3

**Comment by caarlos0** [comment]
goreleaser was built thinking on building binaries only... 

["Yes is forever"](https://twitter.com/solomonstre/status/715277134978113536), so, I rather not do this - or at least not right now.

Thanks for the suggestion either way.

**Comment by qmuntal** [comment]
I understand that this can add complexity to the code base and that would require effort to maintain, but a `dll` file is also a binary, so you are not breaking the foundation goals. In fact, goreleaser can be used to build shared libraries, the only problem is that the binary file name is wrong.

Would you at least consider fixing this issue if I create a PR with a possible solution?

Thanks!

**Comment by caarlos0** [comment]
> I understand that this can add complexity to the code base and that would require effort to maintain, but a `dll` file is also a binary, so you are not breaking the foundation goals. In fact, goreleaser can be used to build shared libraries, the only problem is that the binary file name is wrong.

yeah, you're right about that.


maybe if we add a custom `extension` config field (which defaults to exe on windows)?

**Comment by qmuntal** [comment]
Great! The custom `extension` is a good solution but tricky to define, as the file extension depends on the `goos` + `goarch` combination, and I can't think on an easy way to represent this in yaml (I'm not an expert neither).

What's clear is that the [current defaults](https://github.com/goreleaser/goreleaser/blob/master/internal/pipe/build/build.go#L130) must be maintained, which are:
* goos: `windows` -> `.exe`
* goos: `js`, goarch: `wasm `-> `.wasm`

Another approach could be to extend the list of hard-coded extensions to take into account the `buildmode` in windows:
```go
func extFor(target string, flags string) string {
  if strings.Contains(target, "windows") {
    if strings.Contains(flags, "-buildmode=c-shared") {
      return ".dll"
    } else if strings.Contains(flags, "-buildmode=c-archive") {
      return ".lib"
    }
    return ".exe"
  }
  if target == "js_wasm" {
    return ".wasm"
  }
  return ""
}
```

**Comment by caarlos0** [comment]
sounds good as well..

**Comment by caarlos0** [comment]
fixed in #1243
